pink/tasks/linear_holonomic_task.py

```diff
@@ -0,0 +1,169 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Linear Holonomic task implementation."""
+
+from typing import Sequence, Tuple, Union
+
+import numpy as np
+
+from ..configuration import Configuration
+from .exceptions import TaskJacobianNotSet
+from .task import Task
+
+
+class LinearHolonomicTask(Task):
+    r"""General class for linear holonomic tasks.
+
+    Attributes:
+        A: matrix that relates the following relationship:
+
+        .. math::
+            e(q) = Aq - b
+            \dot{e}(q) := A\dot{q}
+
+        where :math: `e(q) \in \mathbb{R}^{k}` is the quantity that the task 
+            aims to derive to zero (:math:`k` is the dimension of the task).
+        cost: joint angular error cost in 
+            :math:`[\mathrm{cost}] / [\mathrm{rad}]`.
+
+    Note:
+        A linear holonomic task is typically used for a robot 
+        that has mechanical constraint (e.g., closed loop kinematics).
+        Floating base coordinates are not affected by this task.
+    """
+
+    A: np.ndarray
+    cost: Union[float, Sequence[float]]
+
+    def __init__(
+        self, A: np.ndarray, cost: Union[float, Sequence[float]]
+    ) -> None:
+        r"""Create task.
+
+        Args:
+            A: Jacobian matrix of a linear holonomic contraint.
+            cost: joint angular error cost in
+                :math:`[\mathrm{cost}] / [\mathrm{rad}]`.
+        """
+        if isinstance(cost, float):
+            assert A.shape[0] == 1
+        else:
+            assert A.shape[0] == len(cost)
+
+        self.A = A
+        self.cost = cost
+
+    def compute_error(self, configuration: Configuration) -> np.ndarray:
+        r"""Compute the task error function.
+
+        The error function :math:`e(q) \in \mathbb{R}^{k}` is the quantity that
+        the task aims to drive to zero (:math:`k` is the dimension of the
+        task). It appears in the first-order task dynamics:
+
+        .. math::
+
+            J(q) \Delta q = \alpha e(q)
+
+        The Jacobian matrix :math:`J(q) \in \mathbb{R}^{k \times n_v}`,
+        with :math:`n_v` the dimension of the robot's tangent space, is
+        computed by :func:`Task.compute_jacobian`, while the configuration
+        displacement :math:`\\Delta q` is the output of inverse kinematics. The
+        error vector :math:`e(q)` is multiplied by the task gain :math:`\alpha
+        \in [0, 1]`. The gain is usually 1 for dead-beat control (*i.e.*
+        converge as fast as possible), but it can also be lower for some extra
+        low-pass filtering.
+
+        Args:
+            configuration: Robot configuration :math:`q`.
+
+        Returns:
+            Task error vector :math:`e(q)`.
+        """
+        if self.A is None:
+            raise TaskJacobianNotSet("no task Jacobian set")
+
+        return np.zeros(self.A.shape[0])
```

Comment:
The error should $e(q) = A q - b$ but in this current implementation $b=0$.

Can you add a ``self.b`` member and compute the error as ``self.A @ configuration.q - b``?

Reply:
One question here is that we need the b. I thought e(q) was in the tangent space of the configuration, and thus we only need the relationship e(q) = A\dot(q), which leads to e(q) = A \Delta q, where the A matrix is a Jacobian matrix. Am I missing something?

Reply:
Oh, I now understand. Another comment is that since we are dealing with the holonomic task and to be consistent with the configuration space and its tangent space dimension, matrix A should exclude the floating base joint, right?

Reply:
Ah, absolutely, you're right. Having $A$ over configurations generates a dimension mismatch, and I overlooked the fact that we optimize over $v$ in tangent space rather than $\dot{q}$.

We can define things properly with lie theory, but I think that's beyond the scope of this PR, scope that we already extended quite a bit since your original proposal. (Thanks for keeping up! :smiley:) If you agree then, I suggest we keep your current implementation, add some unit tests to make sure everything is consistent, and merge this PR.

Then, in a future PR, we can work on a Lie-consistent version.

Reply:
Sounds good to me!